jax/interpreters/ad.py

```diff
@@ -216,10 +218,10 @@ def is_linear(var):
       cts_in = map(read_cotangent, eqn.outvars)
     else:
       cts_in, = map(read_cotangent, eqn.outvars)
-    if eqn.primitive.call_primitive:
-      call_jaxpr, params = core.extract_call_jaxpr(eqn.primitive, eqn.params)
-      cts_out = get_primitive_transpose(eqn.primitive)(
-          params, call_jaxpr, invals, cts_in)
+    if isinstance(eqn.primitive, (core.CallPrimitive, core.MapPrimitive)):
+      call_jaxpr, params = core.extract_call_jaxpr(eqn.params)
+      transpose_rule = get_primitive_transpose(eqn.primitive)
+      cts_out = transpose_rule(params, call_jaxpr, invals, cts_in)
```

Comment:
Thoughts on whether rules for HOPs (like the transpose rule, but also process_call/process_map) should treat `call_jaxpr` as just another part of `params` now?

Reply:
Oh hmm. An alternative would be to create a special Eqn for these which has a jaxpr slot, right? That seems better...

Reply:
Or, I guess, a special non-params slot on the primitive. Maybe that's what you were thinking too?

Reply:
Ah, that doesn't make sense: there's only one instance of the primitive. Params are really things we bundle with primitives in a jaxpr eqn. So I think we'd need special eqns.

I'll give that a shot!

Reply:
I just meant that George's vision was to make `call_jaxpr` just another param; that would work for transpose and translation (plus any other initial style transforms) but (of course!) not for final style transforms. So I think my comment was mostly off base.

Reply:
Yeah but I think it’s not ideal as a param because then we have some invariants to enforce/check everywhere rather than having those invariants built into the data structure.

Reply:
The reason I thought that putting call_jaxpr in params is that it makes it easier to explain the Jaxpr language to users, since the initial style primitives already have Jaxprs in params. Perhaps we can take all the subjaxprs from params and keep them separate, with only `cond` having two of them?

Reply:
That's a great point.

I talked with @gnecula a bit more this morning and he made the point that we could (and likely should) treat these higher-order primitives more uniformly. In particular, we can treat staged-out xla_call and xla_pmap like the other higher-order primitives and not need a special case in jaxpr interpreters for them. We had some more specific ideas but basically our conclusion was to try out some things and see how they looked.